cmd/rpcdaemon/commands/trace_api.go

```diff
@@ -0,0 +1,246 @@
+}
+
+// TraceAPIImpl is implementation of the TraceAPI interface based on remote Db access
+type TraceAPIImpl struct {
+	db        ethdb.KV
+	dbReader  ethdb.Getter
+	maxTraces uint64
+}
+
+// NewPrivateDebugAPI returns PrivateDebugAPIImpl instance
+func NewTraceAPI(db ethdb.KV, dbReader ethdb.Getter, maxTraces uint64) *TraceAPIImpl {
+	return &TraceAPIImpl{
+		db:        db,
+		dbReader:  dbReader,
+		maxTraces: maxTraces,
+	}
+}
+
+func retrieveHistory(tx ethdb.Tx, addr *common.Address) ([]uint64, error) {
+	addrBytes := addr.Bytes()
+	ca := tx.Cursor(dbutils.AccountsHistoryBucket).Prefix(addrBytes)
+
+	var blockNumbers []uint64
+
+	for k, v, err := ca.First(); k != nil; k, v, err = ca.Next() {
+		if err != nil {
+			return nil, err
+		}
+		numbers, _, err := dbutils.WrapHistoryIndex(v).Decode()
+		if err != nil {
+			return nil, err
+		}
+		blockNumbers = append(blockNumbers, numbers...)
+	}
+	// Remove dublicates
+	return blockNumbers, nil
+}
+
+func isAddressInFilter(addr *common.Address, filter []*common.Address) bool {
+	if filter == nil {
+		return true
+	}
+	for _, f := range filter {
+		if bytes.Equal(addr.Bytes(), f.Bytes()) {
+			return true
+		}
+	}
+	return false
+}
+
+// Filter Implements trace_filter
+func (api *TraceAPIImpl) Filter(ctx context.Context, req TraceFilterRequest, config *eth.TraceConfig) ([]interface{}, error) {
+	var filteredTransactionsHash []common.Hash
+	resp := []interface{}{}
+	var maxTracesCount uint64
+
+	if req.Count == nil {
+		maxTracesCount = api.maxTraces
+	} else {
+		maxTracesCount = *req.Count
+		if req.After != nil {
+			maxTracesCount += *req.After
+		}
+		if maxTracesCount > api.maxTraces {
+			maxTracesCount = api.maxTraces
+		}
+	}
+	if err := api.db.View(ctx, func(tx ethdb.Tx) error {
+		if req.FromAddress != nil || req.ToAddress != nil { // use address history index to retrieve matching transactions
+			if req.FromAddress != nil { // use FromAddress
+				for _, addr := range req.FromAddress {
+
+					addrBytes := addr.Bytes()
+					blockNumbers, err := retrieveHistory(tx, addr)
+					if err != nil {
+						return err
+					}
+					for _, num := range blockNumbers {
+
+						if req.FromBlock != nil && uint64(*req.FromBlock) > num {
+							continue
+						}
+
+						if req.ToBlock != nil && uint64(*req.ToBlock) < num {
+							continue
+						}
+
+						block := rawdb.ReadBlockByNumber(api.dbReader, num)
+						senders := rawdb.ReadSenders(api.dbReader, block.Hash(), num)
+						txs := block.Transactions()
+						for i, tx := range txs {
+							if uint64(len(filteredTransactionsHash)) == maxTracesCount {
+								if uint64(len(filteredTransactionsHash)) == api.maxTraces {
+									return fmt.Errorf("too many traces found")
+								}
+								return nil
+							}
+
+							var to *common.Address
+							if tx.To() == nil {
+								to = &common.Address{}
+							} else {
+								to = tx.To()
+							}
+							if !isAddressInFilter(to, req.ToAddress) {
```

Comment:
Probably easier to express it `if req.ToAddress != nil && *to != *req.ToAddress` then construct such a function

Reply:
`toAddress` is an array not a single value